pkg/querier/queryrange/cache.go

```diff
@@ -0,0 +1,346 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/common/model"
+	"github.com/uber/jaeger-client-go"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/user"
+)
+
+type lokiResultsCache struct {
+	logger   log.Logger
+	cfg      queryrange.ResultsCacheConfig
+	next     queryrange.Handler
+	cache    cache.Cache
+	limits   Limits
+	splitter queryrange.CacheSplitter
+	merger   queryrange.Merger
+}
+
+// NewLokiResultsCacheMiddleware is a cache implementation for "series" and "label" requests
+// this code is mostly taken from cortex's result_cache middleware with only few changes.
+// Main idea to implement custom cache middleware is that with cortex's cache middleware
+// code, we can't correctly cache "series" and "label" responses since they doesn't have a
+// time range associtated with it. This custom code addresses the limitation.
+func NewLokiResultsCacheMiddleware(
+	logger log.Logger,
+	cfg queryrange.ResultsCacheConfig,
+	splitter queryrange.CacheSplitter,
+	limits Limits,
+	merger queryrange.Merger,
+) (queryrange.Middleware, cache.Cache, error) {
+	c, err := cache.New(cfg.CacheConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return &lokiResultsCache{
+			logger:   logger,
+			cfg:      cfg,
+			next:     next,
+			cache:    c,
+			limits:   limits,
+			merger:   merger,
+			splitter: splitter,
+		}
+	}), c, nil
+}
+
+func (s lokiResultsCache) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	var (
+		key      = s.splitter.GenerateCacheKey(userID, r)
+		extents  []queryrange.Extent
+		response queryrange.Response
+	)
+
+	// check if cache freshness value is provided in legacy config
+	maxCacheFreshness := s.cfg.LegacyMaxCacheFreshness
+	if maxCacheFreshness == time.Duration(0) {
+		maxCacheFreshness = s.limits.MaxCacheFreshness(userID)
```

Comment:
I think we want the user limits to override the the configuration parameter. I think you can extend our `WithDefaultLimits` implementation to support `LegacyMaxCacheFreshness` as well:
https://github.com/grafana/loki/blob/master/pkg/querier/queryrange/limits.go#L30-L45